test/exercises/diffusion/heat_2d_dist_buffers.chpl

```diff
@@ -0,0 +1,149 @@
+config const writeTime = false;
+
+// compile with `-sRunCommDiag=true` to see comm diagnostics
+use CommDiagnostics;
+config param RunCommDiag = false;
+
+// declare configurable constants with default values
+config const nx = 256,      // number of grid points in x
+             ny = 256,      // number of grid points in y
+             nt = 50,       // number of time steps
+             alpha = 0.25;  // diffusion constant
+
+// define distributed domains and block-distributed array
+const indices = {0..<nx, 0..<ny},
+      indicesInner = indices.expand(-1),
+      INDICES = Block.createDomain(indices);
+var u: [INDICES] real;
+
+// apply initial conditions
+u = 1.0;
+u[nx/4..nx/2, ny/4..ny/2] = 2.0;
+
+// a type for creating a "skyline" array of halo buffers
+record haloArray {
+  var d: domain(1);
+  var v: [d] real;
+
+  proc init() do this.d = {0..0};
+  proc init(r: range(int)) do
+    this.d = {r};
+}
+
+// set up array of halo buffers over same distribution as 'u.targetLocales'
+var LOCALE_DOM = Block.createDomain(u.targetLocales().domain);
+var haloArrays: [LOCALE_DOM] [0..<4] haloArray;
+
+// buffer edge indices: North, East, South, West
+param N = 0, S = 1, E = 2, W = 3;
+
+// number of tasks per dimension based on Block distribution's decomposition
+const tidXMax = u.targetLocales().dim(0).high,
+      tidYMax = u.targetLocales().dim(1).high;
+
+// barrier for one task per locale
+var b = new barrier(numLocales);
+
+proc main() {
+  if RunCommDiag then startCommDiagnostics();
+
+  // spawn one task for each locale
+  t.start();
+  coforall (loc, (tidX, tidY)) in zip(u.targetLocales(), LOCALE_DOM) {
+    // run initialization and computation on the task for this locale
+    on loc {
+      // initialize halo arrays
+      haloArrays[tidX, tidY][N] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][S] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][E] = new haloArray(u.localSubdomain().dim(0).expand(1));
+      haloArrays[tidX, tidY][W] = new haloArray(u.localSubdomain().dim(0).expand(1));
+
+      // synchronize across tasks
+      b.barrier();
+
+      // run the portion of the FD computation owned by this task
+      work(tidX, tidY);
+    }
+  }
+
+  if RunCommDiag {
+    stopCommDiagnostics();
+    printCommDiagnosticsTable();
+  }
+
+  // print final results
+  const mean = (+ reduce u) / u.size,
+        stdDev = sqrt((+ reduce (u - mean)**2) / u.size);
+  t.stop();
+  writeln(abs(0.222751 - stdDev) < 1e-6);
+  if writeTime then writeln("time: ", t.elapsed());
+}
+
+proc work(tidX: int, tidY: int) {
+  // define domains to describe the indices owned by this task
+  const localIndices = u.localSubdomain(here),
+        localIndicesBuffered = localIndices.expand(1),
+        localIndicesInner = localIndices[indicesInner];
+
+	// declare two local arrays for computation on this locale
+  var uLocal1, uLocal2: [localIndicesBuffered] real = 1;
```

Comment:
Oh wait, you're not operating on the distributed arrays, you're just swapping data in and out of them into these local arrays?  That seems a little unfortunate from a memory usage perspective (`3*n**2` data rather than `2*n**2`).  It also surprises me since I assumed that the creation of the `haloArray` record was to deal with the fact that `u` didn't have buffer space.

As an alternative, should the global array of arrays be an array of these localIndicesBuffered arrays so that each locale can write directly into its neighbors' buffers rather than needing to use the haloArray records as bounce buffers (requiring rank-change slicing and an additional copy).

Why u1 / u2 as names here, but u / un in the original code?  Is there a way to unify names between versions to make them more symmetric?

Reply:
> should the global array of arrays be an array of these localIndicesBuffered arrays so that each locale can write directly into its neighbors' buffers rather than needing to use the haloArray records as bounce buffers (requiring rank-change slicing and an additional copy)

This is what the "exchanges" version of the code does

Reply:
> That seems a little unfortunate from a memory usage perspective (3*n**2 data rather than 2*n**2)

I agree. This results from trying to keep a similar computational flow across versions:
1. create a distributed array and apply initial conditions (`u[nx/4..nx/2, ny/4..ny/2] = 2.0;` in this case).
2. start timing
3. run distributed memory solver (the `coforall` loop over `work` in this case)
4. stop timing
5. compute final results

I can adjust and move (1) and (5) into `work` though, but I think that makes this code quite a bit different than the other versions and probably makes it clunkier to time only the "kernel" execution. Let me know if you agree.

Reply:
Another approach that would preserve the current workflow would be to not allocate the uLocal1/uLocal2 arrays and do the stencil computation directly on the local portions of the original u/un arrays, simply with special-case code for the four boundaries (basically, stripmine those iterations out of the `forall` loop).

Or...  is there a version in which only one temporary local array is required, and the local elements of the distributed array are used as the second buffer?  (downside: array swap is going to be expensive, so you'd probably want to unroll the timestep loop twice, potentially passing the arrays into a helper routine as arguments, to have every other iteration operate on a different array.

As far as suggesting a path forward, I think it depends on what the goal is.  If it's to show people how to write SPMD code in Chapel then I'd probably be inclined to abandon the globally distributed arrays altogether and write it with local arrays, using remote slicing to do the boundary transfers.  If it's to show how to compute on local portions of a distributed array, then I'd probably take one of the two approaches just above.  If we were to stick with the 1.5x memory approach, I think it should be introduced with major caveats, since it limits the size of the problems that can be run.

[this is a big part of why I really don't like using stencils for tutorials... it's not a pattern we've done nearly as much to optimize and tune for as we could]

Reply:
I've refactored the code to operate on the global arrays with special case code for the boundaries.